float64 f0
float64 f1
float64 f2
float64 f3